### PR TITLE
Webhook admission scope

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind api-change
/kind feature

**What this PR does / why we need it**:
Adds the ability to limit the scope of an admission webhook registration to namespaced or cluster-scoped resources.

**Which issue(s) this PR fixes**:
xref https://github.com/kubernetes/enhancements/issues/492
xref https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md#scope

**Does this PR introduce a user-facing change?**:
```release-note
Admission webhooks rules can now limit scope to only match namespaced, or only cluster-scoped resources with a `scope: "Cluster" | "Namespaced" | "*"` field.
```

/sig api-machinery
/cc @mbohlool

### CODE REVIEW

**Comment by liggitt** [comment]
This is implementing the proposal as specified in https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md#scope

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/rules/rules_test.go [hunk 1]
```diff
@@ -299,3 +325,93 @@ func TestResource(t *testing.T) {
...
+				namespacedAttributes("", "v1", "namespaces", "finalize", "ns", admission.Create),
+			),
+			noMatch: attrList(),
+		},
+	}
+	keys := sets.NewString()
+	for name := range table {
+		keys.Insert(name)
+	}
+	for _, name := range keys.List() {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/rules/rules_test.go]: Why can't we directly do `for name, tt := range table`?

File: staging/src/k8s.io/api/admissionregistration/v1beta1/generated.proto [hunk 2]
```diff
@@ -81,6 +81,17 @@ message Rule {
   // Depending on the enclosing object, subresources might not be allowed.
   // Required.
   repeated string resources = 3;
+
+  // Scope specifies the scope of this rule.
+  // Valid values are "Cluster", "Namespaced", and "*"
+  // "Cluster" means that only cluster-scoped resources will match this rule.
+  // Namespace API objects are cluster-scoped.
+  // "Namespaced" means that only namespaced resources will match this rule.
+  // "*" means that there are no scope restrictions.
```
Comment (caesarxuchao) [inline staging/src/k8s.io/api/admissionregistration/v1beta1/generated.proto]: How about explicitly calling out subresources, "subresources inherits the scope of the main resource"?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/rules/rules_test.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/rules/rules_test.go]: this gives a stable order so failing tests don't jump around in order

File: staging/src/k8s.io/api/admissionregistration/v1beta1/generated.proto (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/api/admissionregistration/v1beta1/generated.proto]: updated

**Comment by caesarxuchao** [comment]
lgtm

File: staging/src/k8s.io/api/admissionregistration/v1beta1/types.go [hunk 3]
```diff
@@ -49,8 +49,32 @@ type Rule struct {
 	// Depending on the enclosing object, subresources might not be allowed.
 	// Required.
 	Resources []string `json:"resources,omitempty" protobuf:"bytes,3,rep,name=resources"`
+
+	// Scope specifies the scope of this rule.
```
Comment (smarterclayton) [inline staging/src/k8s.io/api/admissionregistration/v1beta1/types.go]: `scope`

File: staging/src/k8s.io/api/admissionregistration/v1beta1/types.go [hunk 4]
```diff
@@ -49,8 +49,32 @@ type Rule struct {
...
+	// Valid values are "Cluster", "Namespaced", and "*"
+	// "Cluster" means that only cluster-scoped resources will match this rule.
+	// Namespace API objects are cluster-scoped.
+	// "Namespaced" means that only namespaced resources will match this rule.
```
Comment (smarterclayton) [inline staging/src/k8s.io/api/admissionregistration/v1beta1/types.go]: I'm wondering if we should make this `Namespace`.  We call scopes `Cluster` and `Namespace` in other places (technical doc, mostly).  I guess we used `namespaced: true` in APIResource for discovery.  Shame we didn't use `scope: Namespaced` there too.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/rules/rules.go [hunk 5]
```diff
@@ -50,6 +53,22 @@ func exactOrWildcard(items []string, requested string) bool {
...
 
+var namespaceResource = schema.GroupVersionResource{Group: "", Version: "v1", Resource: "namespaces"}
+
+func (r *Matcher) scope() bool {
+	if r.Rule.Scope == nil || *r.Rule.Scope == v1beta1.AllScopes {
+		return true
+	}
+	switch *r.Rule.Scope {
+	case v1beta1.NamespacedScope:
+		return r.Attr.GetResource() != namespaceResource && r.Attr.GetNamespace() != metav1.NamespaceNone
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/rules/rules.go]: Namespace of Namespace should be empty string?  Or is this an admission quirk?  Deserves a comment.

Elsewhere (in resource builder) namespace of namespace is empty.

File: staging/src/k8s.io/api/admissionregistration/v1beta1/types.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/api/admissionregistration/v1beta1/types.go]: CRD scope uses "Cluster" and "Namespaced", was intended to match that

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/rules/rules.go [hunk 6]
```diff
@@ -50,6 +53,22 @@ func exactOrWildcard(items []string, requested string) bool {
...
+	case v1beta1.ClusterScope:
+		return r.Attr.GetResource() == namespaceResource || r.Attr.GetNamespace() == metav1.NamespaceNone
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/rules/rules.go]: deserves a bit more comments I think - nuanced code.  fine to put it on the method.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/rules/rules.go (hunk 5 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/rules/rules.go]: request attributes populate namespace attribute based on path segments, so an individual namespace populates that attribute. will comment on the method

**Comment by liggitt** [comment]
addressed comments

**Comment by liggitt** [comment]
included in docs PR for 1.15 at https://github.com/kubernetes/website/pull/14671

**Pull request merged**